udev_device.c

```diff
@@ -540,6 +561,85 @@ static void set_properties_from_props(struct udev_device *udev_device)
     udev_list_entry_add(&udev_device->properties, "ID_PATH", id, 0);
 }
 
+static int set_properties_from_helper(struct udev_device *udev_device)
+{
+    char *envp[UEVENT_NUM_ENVP];
+    char env[UEVENT_BUFFER_SIZE];
+
+    size_t env_offset = 0;
+    size_t env_idx = 0;
+    struct udev_list_entry *entry;
+
+    for (entry = udev_device_get_properties_list_entry(udev_device);
+         entry;
+         entry = udev_list_entry_get_next(entry)) {
+        const char *name = entry->name;
+        const char *value = entry->value ? entry->value : "";
+
+        // '{name}={value}\0'
+        size_t max_size = strlen(name) + 1 + strlen(value) + 1;
+
+        assert(env_offset + max_size < UEVENT_BUFFER_SIZE);
+        snprintf(&env[env_offset], max_size, "%s=%s", name, value);
+
+        assert(env_idx < UEVENT_NUM_ENVP);
+        envp[env_idx] = &env[env_offset];
+
+        env_offset += max_size;
+        env_idx++;
+    }
+
+    assert(env_idx < UEVENT_NUM_ENVP);
+    envp[env_idx] = NULL;
+
+    int out_pipe[2];
+
+    if (pipe2(out_pipe, O_CLOEXEC)) {
+        return -1;
+    }
+
+    int null_fd = open("/dev/null", O_CLOEXEC | O_RDWR);
+
+    if (null_fd < 0) {
+        return -1;
+    }
+
+    posix_spawn_file_actions_t actions;
+
+    posix_spawn_file_actions_init(&actions);
+    posix_spawn_file_actions_addclose(&actions, out_pipe[0]);
+    posix_spawn_file_actions_adddup2(&actions, null_fd, 0);
+    posix_spawn_file_actions_adddup2(&actions, out_pipe[1], 1);
+    posix_spawn_file_actions_adddup2(&actions, null_fd, 2);
+    posix_spawn_file_actions_addclose(&actions, out_pipe[1]);
+
+    pid_t pid;
+
+    char argv0[] = RULES_HELPER_PATH;
+    char *argv[] = { argv0, NULL };
+
+    if (posix_spawn(&pid, argv0, &actions, NULL, argv, envp)) {
+        return -1;
+    }
+
+    posix_spawn_file_actions_destroy(&actions);
+    close(out_pipe[1]);
+    close(null_fd);
+
+    FILE *file = fdopen(out_pipe[0], "r");
+
+    if (!file) {
+        return -1;
```

Comment:
close out_pipe and waitpid(how to prevent zombie otherwise? posix_spawn does not have POSIX_SPAWN_NOZOMBIE flag like [qnx has](http://www.qnx.com/developers/docs/7.0.0/index.html#com.qnx.doc.neutrino.lib_ref/topic/p/posix_spawn.html))

Reply:
I'm not 100% on my solution. I've never had to think about this properly before. Can you confirm it's correctness?

Reply:
Due to an early return there is a leak of out_pipe and zombie process. Make sure all resources are free'd before returning, e.g:

```c
// fix zombie process leak
int status;
waitpid(pid, &status, 0);

FILE *file = fdopen(out_pipe[0], "r");
if (!file) {
    // fix fd leak
    close(out_pipe[0]);
    return -1;
}

set_properties_from_file(udev_device, file);
fclose(file);
return 0;
```